pkg/paths/path.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package paths
+
+import (
+	"strings"
+)
+
+//
+// To avoid confusion about when paths are encrypted, unencrypted, empty or
+// non existent, we create some wrapper types so that the compiler will complain
+// if someone attempts to use one in the wrong context.
+//
+
+// Unencrypted is an opaque type representing an unencrypted path.
+type Unencrypted struct {
+	raw string
+}
+
+// Encrypted is an opaque type representing an encrypted path.
+type Encrypted struct {
+	raw string
+}
+
+//
+// unencrypted paths
+//
+
+// NewUnencrypted takes a raw unencrypted path and returns it wrapped.
+func NewUnencrypted(raw string) Unencrypted {
+	return Unencrypted{raw: raw}
+}
+
+// Valid returns if the unencrypted path is valid, which is the same as not being empty.
+func (path Unencrypted) Valid() bool {
+	return path.raw != ""
+}
+
+// Raw returns the original raw path for the Unencrypted.
+func (path Unencrypted) Raw() string {
+	return path.raw
+}
+
+// String returns a human readable form of the Unencrypted.
+func (path Unencrypted) String() string {
+	return path.Raw()
+}
+
+// Consume attempts to remove the prefix from the Unencrypted path and
+// reports a boolean indicating if it was able to do so.
+func (path Unencrypted) Consume(prefix Unencrypted) (Unencrypted, bool) {
```

Comment:
Is the prefix supposed to be a complete path component?

e.g. is `a` a prefix of `aa/bb/cc`?

Reply:
In general, yes, `a` is a prefix of `aa/bb/cc`, especially for the case where paths are not using encryption. We'd like to match S3 semantics, if possible.